core/metrics/impl/prometheus/registry_impl.hpp

```diff
@@ -0,0 +1,166 @@
+
+    // create family for metrics type
+    template <typename T>
+    void registerFamily(const std::string &name,
+                        const std::string &help = "",
+                        const std::map<std::string, std::string> &labels = {}) {
+      family_.emplace(
+          name,
+          prometheus::detail::Builder<typename MetricInfo<T>::type>()
+              .Name(name)
+              .Help(help)
+              .Labels(labels)
+              .Register(*registry().get()));
+    }
+
+    template <typename T, typename... Args>
+    T *registerMetric(const std::string &name,
+                      const std::map<std::string, std::string> &labels,
+                      Args... args) {
+      auto &var =
+          dynamic_cast<prometheus::Family<typename MetricInfo<T>::type> &>(
+              family_.at(name).get())
+              .Add(labels, args...);
+      return &std::get<MetricInfo<T>::index>(metrics_).emplace_front(
+          typename MetricInfo<T>::dtype(var));
+    }
+
+    static std::shared_ptr<prometheus::Registry> registry() {
+      static auto registry = std::make_shared<prometheus::Registry>();
+      return registry;
+    }
+
+   public:
+    // Handler has access to internal prometheus registry and gathers metrics,
+    // prepares them for sending by http
+    void setHandler(Handler *handler) override;
+
+    void registerCounterFamily(
+        const std::string &name,
+        const std::string &help,
+        const std::map<std::string, std::string> &labels) override;
+
+    void registerGaugeFamily(
+        const std::string &name,
+        const std::string &help,
+        const std::map<std::string, std::string> &labels) override;
+
+    void registerHistogramFamily(
+        const std::string &name,
+        const std::string &help,
+        const std::map<std::string, std::string> &labels) override;
+
+    void registerSummaryFamily(
+        const std::string &name,
+        const std::string &help,
+        const std::map<std::string, std::string> &labels) override;
+
+    Counter *registerCounterMetric(
+        const std::string &name,
+        const std::map<std::string, std::string> &labels) override;
+
+    Gauge *registerGaugeMetric(
+        const std::string &name,
+        const std::map<std::string, std::string> &labels) override;
+
+    Histogram *registerHistogramMetric(
+        const std::string &name,
+        const std::vector<double> &bucket_boundaries,
+        const std::map<std::string, std::string> &labels) override;
+
+    Summary *registerSummaryMetric(
+        const std::string &name,
+        const std::vector<std::pair<double, double>> &quantiles,
+        std::chrono::milliseconds max_age,
+        int age_buckets,
+        const std::map<std::string, std::string> &labels) override;
+
+    template <typename T>
+    static typename MetricInfo<T>::type *internalMetric(T *metric) {
+      return &dynamic_cast<typename MetricInfo<T>::dtype *>(metric)->m_;
```

Comment:
I hope this could be redone without dynamic_cast

Reply:
actually it is not
it's created for test purposes only